### ISSUE TITLE
systemd-nspawn `--suppress-sync=true` also suppresses fd validation in `fsync()`

### ISSUE THREAD

**Issue by mgorny** [systemd/systemd#34478]
### systemd version the issue has been seen with

256.5

### Used distribution

Gentoo Linux

### Linux kernel version used

6.10.10-gentoo-dist

### CPU architectures issue was seen on

x86_64

### Component

systemd-nspawn

### Unexpected behaviour you saw

`fsync()` (and related calls) returns 0 when passed invalid fd and `--suppress-sync=true` is used.

This is causing CPython's test suite to fail.

### Steps to reproduce the problem

1. Enter a container with `systemd-nspawn --suppress-sync=true`.
2. Compile and run the following program:
    ```c
    #include <stdio.h>
    #include <unistd.h>
    
    int main() {
            printf("fdatasync() = %d\n", fdatasync(-1));
            perror("errno");
            return 0;
    }
    ```

### Additional program output to the terminal or log subsystem illustrating the issue

```sh
fdatasync = 0
errno: Success

(while without --suppress-sync=true, it gives:
fdatasync = -1
errno: Bad file descriptor
)
```

**Comment by yuwata** [comment]
I cannot reproduce the issue. With #34480, the test binary expectedly fails:
```
$ sudo build/systemd-nspawn -xD / --suppress-sync=yes
░ Spawning container thinkpad-x280-f8767b626563378f on /.#machine.26cb0cda4709bf66.
░ Press Ctrl-] three times within 1s to kill container.
root@thinkpad-x280-f8767b626563378f:~# /path/to/build/test-seccomp
(snip)
/* test_seccomp_suppress_sync */
src/test/test-seccomp.c:1248: Assertion failed: expected "fdatasync(-1)" to fail with error "Bad file descriptor", but it succeeded
seccomp_suppress_sync terminated by signal ABRT.
Assertion 'wait_for_terminate_and_check("seccomp_suppress_sync", pid, WAIT_LOG) == EXIT_SUCCESS' failed at src/test/test-seccomp.c:1261, function test_seccomp_suppress_sync(). Aborting.
Aborted (core dumped)
```

**Comment by yuwata** [comment]
Is your systemd-nspawn built with SECCOMP support??

**Comment by YHNdnzj** [comment]
> I cannot reproduce the issue. With #34480, the test binary expectedly fails:
> ```
> $ sudo build/systemd-nspawn -xD / --suppress-sync=yes
> ░ Spawning container thinkpad-x280-f8767b626563378f on /.#machine.26cb0cda4709bf66.
> ░ Press Ctrl-] three times within 1s to kill container.
> root@thinkpad-x280-f8767b626563378f:~# /path/to/build/test-seccomp
> (snip)
> /* test_seccomp_suppress_sync */
> src/test/test-seccomp.c:1248: Assertion failed: expected "fdatasync(-1)" to fail with error "Bad file descriptor", but it succeeded
> seccomp_suppress_sync terminated by signal ABRT.
> Assertion 'wait_for_terminate_and_check("seccomp_suppress_sync", pid, WAIT_LOG) == EXIT_SUCCESS' failed at src/test/test-seccomp.c:1261, function test_seccomp_suppress_sync(). Aborting.
> Aborted (core dumped)
> ```
> Of course, the test succeeds when --suppress-sync=no.

Hmm? That's exactly the issue here, no? The error clearly shows that fdatasync() unexpectedly succeeded.

**Comment by yuwata** [comment]
> Hmm? That's exactly the issue here, no? The error clearly shows that fdatasync() unexpectedly succeeded.

When `--suppress-sync=yes`, `fdatasync(-1)` should succeed, and the test case added in #34480 will fail, as the test assumes that no seccomp filter for sync related syscall are installed and expects the call fails.
When `--suppress-sync=no`, `fdstasync(-1)` should fail, and the test case will succeed as expected.

**Comment by yuwata** [comment]
Anyway, please check the below at first:
> Please check `systemd-nspawn --version` shows `+SECCOMP`.

**Comment by mgorny** [comment]
> Anyway, please check the below at first:
> 
> > Please check `systemd-nspawn --version` shows `+SECCOMP`.

Yes, it does.

```
systemd 256 (256.5)
+PAM +AUDIT -SELINUX -APPARMOR +IMA +SMACK +SECCOMP +GCRYPT +GNUTLS +OPENSSL +ACL +BLKID +CURL +ELFUTILS -FIDO2 +IDN2 -IDN -IPTC +KMOD -LIBCRYPTSETUP -LIBCRYPTSETUP_PLUGINS +LIBFDISK +PCRE2 -PWQUALITY -P11KIT +QRENCODE +TPM2 +BZIP2 +LZ4 +XZ +ZLIB +ZSTD -BPF_FRAMEWORK -XKBCOMMON +UTMP -SYSVINIT +LIBARCHIVE
```

**Comment by YHNdnzj** [comment]
> When --suppress-sync=yes, fdatasync(-1) should succeed

Again, this is what the whole report is about...

**Comment by yuwata** [comment]
Ah, I misunderstood the report.

The behavior is expected. `--suppress-sync=yes` installs seccomp filter for fsync and friends to make they unconditionally succeed. So, even an invalid fd is passed, it succeeds.

**Comment by poettering** [comment]
I think the original reporter has a point: to minimize unexpected breakages by --suppress-sync= we should probably extend our seccomp filter to return EBADF on negative fds. I mean, the goal with that option is to just speed up things, but from app pov it shouldn't change behaviour as much as possible.

i mean, everything has boundaries: tweaking the seccomp filter to refuse negative fds should be fairly easy. But tweaking it to refuse positive-but-invalid fds would be too much to ask, and simply not implementable via seccomp. But just because we cannot do the latter, we still should do the former.

**Comment by yuwata** [comment]
OK. Fix is waiting in #34480.

**Comment by mgorny** [comment]
Thank you!
